common/src/address/mod.rs

```diff
@@ -55,80 +41,72 @@ impl From<Bech32Error> for AddressError {
     }
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Encode)]
-pub struct Address {
+#[must_use]
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Address<T> {
     address: String,
+    _marker: std::marker::PhantomData<T>,
 }
 
-impl Address {
-    pub fn new_from_destination(
-        cfg: &ChainConfig,
-        destination: &Destination,
-    ) -> Result<Self, AddressError> {
-        Self::new_with_hrp(cfg.address_prefix(destination), destination.encode())
+impl<T> Address<T> {
+    pub fn get(&self) -> &str {
+        &self.address
+    }
+}
+
+impl<T: Addressable> Address<T> {
+    pub fn new(cfg: &ChainConfig, object: &T) -> Result<Self, AddressError> {
+        Self::new_with_hrp(
+            T::address_prefix(object, cfg),
+            object.encode_to_bytes_for_address(),
+        )
     }
 
-    pub(crate) fn new_with_hrp<T: AsRef<[u8]>>(hrp: &str, data: T) -> Result<Self, AddressError> {
+    fn new_with_hrp<D: AsRef<[u8]>>(hrp: &str, data: D) -> Result<Self, AddressError> {
         let d = encoding::encode(hrp, data)?;
-        Ok(Self { address: d })
+        Ok(Self {
+            address: d,
+            _marker: std::marker::PhantomData,
+        })
     }
 
-    pub fn destination(&self, cfg: &ChainConfig) -> Result<Destination, AddressError> {
+    pub fn decode_object(&self, cfg: &ChainConfig) -> Result<T, AddressError> {
         let data = encoding::decode(&self.address)?;
-        let raw_dest = data.data();
-        let destination = Destination::decode_all(&mut &raw_dest[..])
-            .map_err(|e| AddressError::DestinationDecodingError(e.to_string()))?;
-        if data.hrp() != cfg.address_prefix(&destination) {
+        let raw_data = data.data();
+        let result = T::decode_from_bytes_from_address(raw_data)
```

Comment:
Reads a bit weird. Maybe `decode_from_address_bytes()`?

Reply:
I want to avoid saying "address bytes" because it can be interpreted as serialization of the address itself.